activesupport/lib/active_support/abstract_class.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module AbstractClass
+    extend ActiveSupport::Concern
+
+    included do
+      if defined?(@root_class)
+        raise AbstractClassError, "#{self} is already part of #{root_class}'s hierarchy"
+      else
+        @root_class = self
+      end
+    end
+
+    class_methods do
+      # Returns the class descending directly from the root class, or
+      # an abstract class, if any, in the inheritance hierarchy.
+      #
+      # If A extends the root class, A.base_class will return A. If B descends from A
+      # through some arbitrarily deep hierarchy, B.base_class will return A.
+      #
+      # If B < A and C < B and if A is an abstract_class then both B.base_class
+      # and C.base_class would return B as the answer since A is an abstract_class.
+      def base_class
+        unless self < root_class
+          # TODO: Use a different error class. Can we do this without breaking backwards compatibility?
+          raise ActiveRecordError, "#{name} doesn't belong in a hierarchy descending from #{root_class}"
```

Comment:
As a separate concern in `ActiveSupport`, we couldn't want to `raise ActiveRecordError`. However, perhaps we still need this for backwards compatability?

If so, we could raise something like `ActiveSupportError` here, and then have something like this in `ActiveRecord::Inheritance`

```ruby

def base_class
  super
rescue ActiveSupportError => error
  raise ActiveRecordError, error.message
end
```